RFC: Add the `group_by` and `group_by_mut` methods to slice

Kerollmops opened the pull request:
This RFC propose to add two new methods to the slice, the `group_by` and `group_by_mut`. These two will provide a way to iterate over non-overlapping sub-slices of a base slice that are separated by the predicate given by the user (e.g. `Partial::eq`, `|a, b| a < b`).

The predicate is called on two elements following themselves, it means the predicate is called on `slice[0]` and `slice[1]` then on `slice[1]` and `slice[2]`...

```rust
let slice = &[1, 1, 1, 3, 3, 2, 2, 2];

let mut iter = slice.group_by(|a, b| a == b);

assert_eq!(iter.next(), Some(&[1, 1, 1][..]));
assert_eq!(iter.next(), Some(&[3, 3][..]));
assert_eq!(iter.next(), Some(&[2, 2, 2][..]));
assert_eq!(iter.next(), None);
```

[Pending Pull Request](https://github.com/rust-lang/rust/pull/51606)

[Work around temporary library](https://github.com/Kerollmops/slice-group-by)

[Rendered](https://github.com/Kerollmops/rust-rfcs/blob/group-by/text/0000-group-by.md)

CodesInChaos commented:
I don't like the naming. Something like `split_between` would fit the behaviour better.

Lokathor commented:
I guess, but this is exactly what [groupBy](https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-List.html#v:groupBy) does in Haskell, so the name should stay.

Kerollmops commented:
The function you describe is the same as the [GroupWith](http://hackage.haskell.org/package/base-4.11.1.0/docs/GHC-Exts.html#v:groupWith) behavior in Haskell.

ssokolow commented:
> I don't like the naming. group by in SQL and C# work quite differently from what you propose, they map each item to a key and then group all items having the same key.
> 
> Something like split_between would fit the behaviour better.

> I guess, but this is exactly what groupBy does in Haskell, so the name should stay.

I don't like the "**My** language is more important than yours" tone of this exchange.

That said, from my perspective, a name like `split_between` still wins out because I consider "no preconceptions" to be better for learnability than being torn between the incompatible preconceptions of two different languages, both significant.

Assuming that `group_by` turns out to be a non-starter for this reason, is there another name with prior art that could be proposed?

rvolgers commented:
This seems pretty similar to the existing `take_while` and `skip_while` functions. But it's even more specific, so I'd want to see some clear use cases. Maybe it's obvious to Haskell programmers, but it isn't to me :smile:

I can imagine some cases where you iterate over a list and you can skip some work if some predicate hasn't changed. But in that case, it seems clearer to me to write it as a single loop with that part of the behavior written out with a `prev_value` variable and in imperative form. It may be a couple more lines of code but it makes the shape of the code a lot more representative of what is being accomplished (a single iteration where the processing has a special fast path case).

Centril commented:
@ssokolow 

> I don't like the "My language is more important than yours" tone of this exchange.

I don't think that's it. There's a good reason to favor a Haskell based name in the case of iterator methods, functions, and related things because Rust already uses Haskell based naming in such cases. For example: `map`, `zip`, `fold`, `take`, `take_while`, `scan`, `partition`, `all`, `any`, `find`, `unzip`, `cycle`, `sum`, `product`, `repeat`. There should be some consistency in naming, and `group_by` is consistent.

There's also [precedent from `itertools`](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.group_by) in naming this operation `group_by`.

I also think it's a telling and descriptive name. You are grouping elements by where the predicate applies. (Yes, a split happens every time the predicate doesn't match, so the order matters, but if you sort, then it does not.)

@rvolgers 

>  But it's even more specific, so I'd want to see some clear use cases. Maybe it's obvious to Haskell programmers, but it isn't to me 😄

Grouping is a pretty common operation; Think of `GROUP BY` in SQL -- a lot of those cases can be expressed as `.group_by(..)`.

> It may be a couple more lines of code but it makes the shape of the code a lot more representative of what is being accomplished (a single iteration where the processing has a special fast path case).

Generally speaking, at least to me, iterator style composition makes data flow clearer :)

leonardo-m commented:
A related by different function in D language:
https://dlang.org/phobos/std_algorithm_iteration.html#group

And Python:
https://docs.python.org/2/library/itertools.html#itertools.groupby

So what semantics do we prefer?

Centril commented:
@leonardo-m It seems to me that the semantics of D and Python can be recovered from the proposed semantics in this RFC (which is the same as the semantics in the itertools crate) so I would say that the proposed solution is more general.

Ixrec commented:
To me, it's not even sufficiently clear what the RFC is proposing. The name `group_by` makes me expect an array like `[1, 1, 2, 1]` to get turned into the groups `[1, 1, 1]` and `[2]`. But the fact that this is an iterator method makes me expect a more linear-time-y behavior like `[1, 1]`, `[2]` then `[1]`. The RFC's summary sentence does use the term "runs", which I am just barely familiar with so that also seems to imply `[1, 1]`, `[2]` then `[1]`. But the RFC doesn't contain a single example of expected inputs and outputs. And in the ~200 lines of tests in the linked implementation, not one test covers this distinction!

The reference implementation of `next()` seems like it would in fact produce `[1, 1]`, `[2]` then `[1]`, so I assume that was the author's intent. But still, that should be made utterly and idiot-proof-ly obvious in the RFC, without having to read an implementation.

Another problem with providing an implementation and no specification is that I have no idea what the guarantees/requirements are for the predicate. Maybe this is a silly question, but am I guaranteed that it gets invoked with (a[0], a[1]) then (a[1], a[2]) and so on in that order exactly once each, or is the implementation allowed to do things like (a[0], a[1]) then (a[0], a[2]) and so on? I assume everyone's assuming the former, but since there isn't a single example or test of a predicate other than equality (`panic_param_ord` doesn't count), it's not super clear what more interesting predicates I should or should not be able to pass to this algorithm.1

Which is all a really long way of saying I don't think we can properly bikeshed the name quite yet. In particular, I think I need examples of more interesting predicates to figure out whether "groups", "runs" or "splitting" is the least misleading term for what's going on.

phaazon commented:
> Assuming that group_by turns out to be a non-starter for this reason, is there another name with prior art that could be proposed?

`regroup_by`?

Centril commented:
@Ixrec I would expect the behavior to be equivalent to https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.group_by except that you get a slice as the element of the returned iterator instead of an `IntoIterator`.

@phaazon What's the prior art for `regroup_by`?

@Kerollmops  Small library additions such as the one proposed in this RFC have historically been accepted with a PR against `rust-lang/rust`, so I would recommend filing a PR against that repo so that we can see the concrete implementation, tests, etc. and then we can have a much clearer idea.

phaazon commented:
@Centril About `regroup_by`, none, and I think:

  - `regroup_by` semantics would sort upfront, which is not the case of this suggested implementation.
  - `group_by` is fine enough but if people don’t like it, it shouldn’t be very hard to find a proper name based on the current `Iterator` other methods.

phaazon commented:
`continuous_chunks` and `continuous_chunks_by`?

Ixrec commented:
@Centril Interesting, the itertools method expects an `element -> key` callback and strongly implies the keys will be compared via PartialEq to determine the groups.

What other use cases does this proposal hope to support by generalizing to a `(a, b) -> bool` callback? If equality and key equality are all we know of, I'd rather have the more obvious itertools signature.

phaazon commented:
(Also, not related to the actual discussion, this is more meta about the commit: @Kerollmops, you seem to have pushed this commit with a company email address. I think you should check twice and possibly rebase with your OSS identity – your commit is also not verified).

Centril commented:
@Ixrec I would provide the following hierarchy:
+ The general operation [`group_by`](http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-List.html#v:groupBy) using `(T, T) -> bool`
+ The half-general operation `group_by_key` or `group_key` using `T -> Key, Key: PartialEq`
+ The specialized operation [`group`](http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-List.html#v:group) using `T: PartialEq`

The two latter ones can be implemented in terms of the first one. I would not use the name `group_by` for `T -> Key, Key: PartialEq` to leave room for the most general version even if we don't add it now.

Kerollmops commented:
It could probably be named `split_by`.

Centril commented:
I think it would be confusing to have [`split`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.split) and then `split_by` with entirely different semantics.

Lokathor commented:
Yeah, split is absolutely not the verb, even if group is also not the verb (though i think group is the best verb here)

uberjay commented:
Ruby calls this chunk — https://ruby-doc.org/core-2.5.1/Enumerable.html#method-i-chunk

Of course, that’s already taken. Would it be too confusing if it were `chunk_by`? Perhaps...

`group_by` makes me think it’ll behave like the (obviously more expensive) ruby method of the same name, which enumerates and groups/collects (up front) into a hash.

Ixrec commented:
It seems like I'm the only one that finds it unintuitive for `group` to only produce contiguous runs of equivalent elements rather than gathering up all equivalent elements. But just in case I'm not alone: `runs`, `runs_by_key` and `runs_by` seem like acceptable names to me.

uberjay commented:
@lxrec - I also find `group_by` unintuitive. It's possible my view is colored by ruby use, but my feeling is that _group_ doesn't convey the sequential nature of the operation.

For fun, I asked someone nearby who isn't familiar with any of the languages in question, and they felt pretty strongly that `chunk` was a far better word than `group`, for this operation. Totally un-scientific survey of one person! 😬 

Now, there's already `chunks`, which is for fixed length chunks -- maybe `chunks_by` wouldn't be that odd? `seq_group_by`?

ssokolow commented:
I'd go with the `runs`-based naming because "run" is existing jargon (See, for example, [Run-length encoding](https://en.wikipedia.org/wiki/Run-length_encoding)) and, if you're familiar with that jargon, it's intuitive for `runs_by` to have a similar meaning to passing a key-derivation function into a sorting routine.

Lokathor commented:
Actually `group_by` _will_ produce all equivalent elements in each output batch if (and only if) the slice is already sorted and the closure used is an equality test. Otherwise you'll see some other sort of result.

Remember that `group_by` takes `&self` (for some `[T]`), so you can't re-arrange any of the values (since you can't mutate the slice's memory backing), and you also can't move the values into temporary memory and sort there (since you don't bound on `T: Copy`). Thus, the only possible output is a series of sub-slices of the starting slice that don't have any order changes.

The only question left is one of the ones you already said: assuming the first slice starts with (a[0], a[1]), do we keep checking the next element against the start of the slice or against the latest element of the slice?

uberjay commented:
Ok, a few interesting alternatives from the thesaurus:

`segment_by` - this actually reads quite nicely to me.
`snip_by` - ok, but a little weird, and probably too similar to `split`.

ssokolow commented:
@uberjay 

I'd go with `segment_by`. `snip_by` has a "too clever" feel to it.

(Aside from having no precedent I'm aware of, it has a feeling of being too focused on process rather than effect. "Snipping" is what you do to accomplish a goal like "trimming" or "splitting" rather than a goal in itself.)

scottmcm commented:
The motivation here reads to me like "this makes it easier to do what this does", which I don't find persuasive.  Why is it common to have this problem, to have the input in exactly the form needed for this proposed implementation, and to not just want the eager `HashMap`-producing version?

@Ixrec You're definitely not alone; I would absolutely expect "group by" to have the relational algebra meaning as well.

Lokathor commented:
well this works in a no_alloc situation, for one

burdges commented:
I suppose `split` and `split_mut` excluding the matched element was a dubious design decision, given that one can shorten slices easily enough.  Anyways you can implement a version that gives slightly short slices like:
```
let mut b = None;
slice.split(|a| { let c = b; b = Some(a);  b != c })
```

Ixrec commented:
I think `segment_by` is my new favorite. It's a better-known term than "runs" _and_ it has a stronger implication of contiguity (contiguousness?) than "group" does.

mark-i-m commented:
I can't say how often I have wanted this feature! It would be especially useful when iterating over file contents.

I think `chunk_by` or `split_by` sound reasonable.

mark-i-m commented:
The only problem with "window" or "chunk" is that they could be interpreted to mean all the chunks are the same size...

Centril commented:
@Ixrec I still prefer `group_by` for consistency reasons; but `segment_by` is not so bad.

joshtriplett commented:
I'd prefer `group_by` as well, largely because it already goes by that name in itertools.

For that matter, I'd like to have a `group` (without the `_by`) that uses `PartialEq`.

Kerollmops commented:
The [itertools `group_by` method](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.group_by) does not work like the one I propose, it asks for a key to map the element and does not return contiguous elements.

Kerollmops commented:
@Centril, as [you propose above](https://github.com/rust-lang/rfcs/pull/2477#issuecomment-397743028), I have done [the work to add](https://github.com/rust-lang/rust/compare/master...Kerollmops:group-by) to the `rust-lang/rust` repo.

The problem is that it will not compile if I do not specify an issue on the unstable stuff but there is no traking issue for the moment, so I put "42" as a placeholder.

Centril commented:
@Kerollmops Leaving some preliminary notes here on your work (ill leave more as inline comments later when you file the PR):

+ You can use `0` as the issue number until a real tracking issue is made.
+ I'd make the feature name a bit more specific: `slice_group_by`.
+ All `unsafe { .. }` elimination blocks should have comments explaining why it is safe.

kennytm commented:
Why everyone immediately starts bikeshedding the name 😕 

I agreed with https://github.com/rust-lang/rfcs/pull/2477#issuecomment-397774952 that the RFC doesn't demonstrate why `group_by` needs to be a part of std, instead of using the existing [`Itertools::group_by`](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.group_by).

Furthermore, checking [sourcegraph](https://sourcegraph.com/search?q=repogroup:crates+%5Cbgroup_by%5Cb+-repo:%5Egithub%5C.com/bluss/rust-itertools%24+-repo:%5Egithub%5C.com/diesel-rs/diesel%24), after removing the repositories `itertools` (which defined `group_by`) and `diesel` (which is a false positive usage, referring to SQL's `GROUP BY` clause), we find that [*only* `clippy`](https://github.com/rust-lang-nursery/rust-clippy/pull/2672) is using this function, suggesting this function isn't in high demand.

You almost always need to sort the slice before using `group_by` to get some meaningful result (similar to `Vec::dedup`). This makes it less convenient to use:

```rust
let mut packages = metadata.packages;
packages.sort_by(|a, b| a.name.cmp(&b.name));    // <------
for group in packages.group_by(|a, b| a.name == b.name) {
    let name = &group[0].name;
    if group.len() > 1 { ... }
}
```

The Map-based refactoring isn't that bad compared with this.

```rust
let mut grouped_packages: HashMap<_, Vec<_>> = HashMap::new();
for package in &metadata.packages {
    grouped_packages.entry(&package.name).or_default().push(package);
}
for (name, group) in grouped_packages {
    if group.len() > 1 { ... }
}
```

phaazon commented:
Now I start to get why @kennytm and @scottmcm feel this way. The RFC should be modified – especially its [motivation](https://github.com/Kerollmops/rust-rfcs/blob/group-by/text/0000-group-by.md#motivation) – to explain why this should be merged to `std` instead of a third-party library, like `itertools`.

Kerollmops commented:
I updated the [motivation section](https://github.com/Kerollmops/rust-rfcs/blob/group-by/text/0000-group-by.md#motivation) as requested.

phaazon commented on text/0000-group-by.md:
```diff
@@ -0,0 +1,101 @@
+Adding this `Iterator` to the standard library will help people split slices by using a custom predicate!
+This `Iterator` is implemented on generic slices to provide performances and flexibility, `GroupBy` implements `DoubleEndedIterator` without any overhead and it does not need any allocation.
+
+There is a similar method that already exists in [the standard library called `split`](https://doc.rust-lang.org/std/primitive.slice.html#method.split) but it will remove the element that does the separation.
```
You still haven’t explained *why* this should be added to `std` / `core` instead of `itertools`. If `itertools` must be fixed or augmented with a new function, this function belongs to `itertools`, not `std`.

Kerollmops replied:
I updated the motivation section.

Kerollmops commented:
I have multiple usecases in one of my project and I wrote custom loops just as temporary fixes.

I do not think that this method fit in the `itertools` library, as the`itertools` description say: _Extra iterator adaptors, functions and macros._. And this function is really optimized for slices/contiguous data.

burdges commented:
Arguably, these should be called `split` and `split_mut` and the existing `split` and `split_mut` should be renamed to `lines_by` or `separated` or whatever, but presumably that's impossible now.  

You can almost implement `split` with this, assuming you capture the closure's lifetimes.
```
pub fn split<'a F>(&'a self, pred: F) -> impl Iterator<Item=&'a [T]>
  where F: FnMut(&T) -> bool + 'a {
    self.group_by(|a,_| pred(a)).map(|x| x.split_last()).1
}
pub fn split_mut<'a,F>(&'a self, pred: F) -> impl Iterator<Item=&'a mut [T]>
  where F: FnMut(&T) -> bool + 'a {
    self.group_by_mut(|a,_| pred(a)).map(|x| x.split_last_mut()).1
}
```
It doesn't quite work however since `pred(self[self.len()-1])` never gets evaluates and `pred` cannot be used in the `map`.  `#[async]` can express this easily though I guess.

Kerollmops commented:
```rust
```

Kerollmops commented:
I think something like this can correctly emulated the current `slice::split` method.

```rust
pub fn split<F>(&self, pred: F) -> impl Iterator<Item=&[T]>
where F: FnMut(&T) -> bool
{
    self.group_by(|a, _| pred(a))
        .enumerate()
        .map(|(i, slice)| {
            match i {
                0 => slice,
                _ => slice[1..],
            }
        })
}
```

But this is not really the subject here, `split` will be kept as you say, the name for the method I propose will be something between `group_by` and `segment_by`.

For the real discussion here is: do we really want to add this method to the standard library ?
@mark-i-m, @Centril and me are likely to merge it in because we faced use cases.

burdges commented:
I think our discussion of deriving `split` from `group_by` reflects positively on including it in the standard library, well `group_by` sounds more fundamental. 

As I said, it's unfortunate `split` looks soo similar to `split_at` while behaving so radically different. :( 

Afaik, you'd need the lifetime so the return can borrow both `self` and `pred`, but you'd also need `pred` to appear in both closures, so you must wrap `pred` in a `RefCell` or `Arc` making it look more like:
```
pub fn split<'a F>(&'a self, pred: F) -> impl Iterator<Item=&'a [T]>
  where F: FnMut(&T) -> bool + 'a {
    let p = RefCell::new(pred);
    self.group_by(|a,_| p.borrow_mut()(a)).map(|x| {
        { let (h,t) = x.split_last();  if pred(*h) { return t; } }
        x
    })
}
pub fn split_mut<'a,F>(&'a self, pred: F) -> impl Iterator<Item=&'a mut [T]>
  where F: FnMut(&T) -> bool + 'a {
    let p = RefCell::new(pred);
    self.group_by_mut(|a,_| p.borrow_mut()(a)).map(|x| {
        { let (h,t) = x.split_last_mut();  if pred(*h) { return t; } }
        x
    })
}
```
In principle, the new `#[async]` generator function would permit borrowing `pred` across a yield point, but of course that's not enough to pass it to two separate closures, so made a mistake there.

uberjay commented:
Apologies about the immediate bike-shedding.

In terms of the actual functionality, I've found the corresponding functionality pretty useful in Ruby-land for segmenting (or chunking, or grouping, if you will :) line-based output into logical chunks more easily. It's not a _great_ day when you're stuck parsing the output of some random command line tool, but `Enumerable#chunk` made it a bit more pleasant.

So, it's been useful for me in situations where sorting beforehand doesn't make sense.

Kerollmops commented:
Ok so the only blocking thing about this RFC seems to be the name chosen (`group_by`), as [@Lokathor says](https://github.com/rust-lang/rfcs/pull/2477#issuecomment-397662688) this is the function name that is used in Haskell, and the Rust standard library is mostly inspired by it.

[@uberjay propose `segment_by`](https://github.com/rust-lang/rfcs/pull/2477#issuecomment-397770964) and I think nobody is opposed to this name.

It seems that the only way I can make this RFC move forward is by changing the name from `group_by` into `segment_by`.

I am not convinced about this renaming, as [@mark-i-m says](https://github.com/rust-lang/rfcs/pull/2477#issuecomment-398226307):
> I think the immediate bikeshedding is a sign that people don't have any larger issues with the proposal...

So I propose to tag this RFC as [`proposed-final-comment-period`](https://github.com/rust-lang/rfcs/labels/proposed-final-comment-period), publish it in [this-week-in-rust](http://this-week-in-rust.org) to make it more visible and see other oppositions than naming.

Kerollmops commented:
@Centril what do we do in case of blocked RFCs ?

Kerollmops commented:
Ping !

uberjay commented:
Totally agree -- it'd be great to have this, regardless of what it's named! I've run across a couple times where it would have come in handy over the past couple months, even!

Kerollmops commented:
As the RFC seems to get stuck I will ping someone from the Library team.
I choose @Kimundi and @sfackler as he has already seen [the final pull request](https://github.com/rust-lang/rust/pull/51606).

Kerollmops commented:
I propose to rename the method `group` and the returned iterator `Group`, this way it will follow the same rules `split` familly follows. [The `slice::split` method returns a `Split` struct][split]. Same pattern for [`rsplit`], [`splitn`], [`rsplitn`] and all the mutable versions.

Note that the `Split` struct is generic on the function so the `Group` struct must not be based on partial equallity. In other word, no `GroupBy`, `GroupByKey` struct, only the `Group` one which is invoqued by the `slice::group` method with a generic function.

[split]: https://doc.rust-lang.org/std/primitive.slice.html#method.split
[`rsplit`]: https://doc.rust-lang.org/std/primitive.slice.html#method.rsplit
[`splitn`]: https://doc.rust-lang.org/std/primitive.slice.html#method.splitn
[`rsplitn`]: https://doc.rust-lang.org/std/primitive.slice.html#method.rsplitn

joshtriplett commented:
Nominating for discussion in the next libs team meeting, to get the process un-stuck after the various bikeshed-painting. :)

Kerollmops commented:
@joshtriplett When does the next libs team meeting occurs ?

Kerollmops commented:
For those interrested, [I have made a temporary library](https://github.com/Kerollmops/group-by) that I will not publish to crates.io providing a workaround. It compiles on stable rust, the version that will be merged will use unstable functions to improve performance and clarity (i.e. [offset_from](https://github.com/Kerollmops/group-by/blob/cab857bae01463dbd0edb99b0e0d7f3624e6c6f5/src/lib.rs#L22-L28)).

```rust
let slice = &[1, 1, 1, 1, 2, 2, 2, 2];

let mut iter = GroupBy::new(slice, |a, b| a == b);

assert_eq!(iter.next(), Some(&[1, 1, 1, 1][..]));
assert_eq!(iter.next(), Some(&[2, 2, 2, 2][..]));
assert_eq!(iter.next(), None);
```

KodrAus commented:
Sorry we dropped the ball on this one. We've actually just recently restarted Libs meetings, so now, two years later we can say that for API additions like this we actually don't typically need an RFC and will happily land an unstable implementation. That wasn't widely understood by the wider Rust org at the time though.

If anybody would like to resubmit https://github.com/rust-lang/rust/pull/51606 we'll be happy to land this API as unstable.

Sorry for leaving this for so long and thank you for putting all this together @Kerollmops 🙇

Kerollmops commented:
Thank you @KodrAus for your time, I understand that this is very hard to follow all of these PRs and RFCs so don't worry, [I created my own library](https://lib.rs/slice-group-by) to support much more than just sequential groups (i.e. linear, binary and, exponential group of search).

[I opened a new PR for this addition](https://github.com/rust-lang/rust/pull/79895).

CodesInChaos commented:
It seems to be that people are discussing two different but related functions here:

1. A function which takes a element->key mapping and compares adjacent keys with `PartialEq`. This is what itertools calls `group_by`
2. A function which takes a split predicate (element, element) -> bool which decides if it should split between those elements. This is what the OP proposes. I do not think this function should be called `group_by`.

RustyYato commented:
Would `split_if` be a better name?

KodrAus commented:
This is now being tracked in https://github.com/rust-lang/rust/issues/80552

@CodesInChaos I've captured your concern about naming in that issue. We can continue discussion over there!

Thanks everybody who contributed here.
